swiftnav/src/coords.rs

```diff
@@ -417,8 +522,102 @@ impl Default for AzimuthElevation {
     }
 }
 
+/// Complete coordinate used for transforming between reference frames
+///
+/// Velocities are optional, but when present they will be transformed
+#[derive(Debug, PartialEq, PartialOrd, Clone)]
+pub struct Coordinate {
+    reference_frame: ReferenceFrame,
+    position: ECEF,
+    velocity: Option<ECEF>,
+    epoch: GpsTime,
+}
+
+impl Coordinate {
+    pub fn new(
+        reference_frame: ReferenceFrame,
+        position: ECEF,
+        velocity: Option<ECEF>,
+        epoch: GpsTime,
+    ) -> Self {
+        Coordinate {
+            reference_frame,
+            position,
+            velocity,
+            epoch,
+        }
+    }
+
+    pub fn without_velocity(
+        reference_frame: ReferenceFrame,
+        position: ECEF,
+        epoch: GpsTime,
+    ) -> Self {
+        Coordinate {
+            reference_frame,
+            position,
+            velocity: None,
+            epoch,
+        }
+    }
+
+    pub fn with_velocity(
+        reference_frame: ReferenceFrame,
+        position: ECEF,
+        velocity: ECEF,
+        epoch: GpsTime,
+    ) -> Self {
+        Coordinate {
+            reference_frame,
+            position,
+            velocity: Some(velocity),
+            epoch,
+        }
+    }
+
+    pub fn reference_frame(&self) -> ReferenceFrame {
+        self.reference_frame
+    }
+
+    pub fn position(&self) -> ECEF {
+        self.position
+    }
+
+    pub fn velocity(&self) -> Option<ECEF> {
+        self.velocity
+    }
+
+    pub fn epoch(&self) -> GpsTime {
+        self.epoch
+    }
+
+    /// Use the velocity term to adjust the epoch of the coordinate.
+    /// When a coordinate has no velocity the position won't be changed.
+    pub fn adjust_epoch(&self, new_epoch: &GpsTime) -> Self {
+        let dt =
+            new_epoch.to_fractional_year_hardcoded() - self.epoch.to_fractional_year_hardcoded();
+        let v = self.velocity.unwrap_or_default();
+
+        Coordinate {
+            position: self.position + dt * v,
+            velocity: self.velocity,
+            epoch: *new_epoch,
+            reference_frame: self.reference_frame,
+        }
```

Comment:
The style I typically write Rust code with definitely has more copies than is strictly necessary, and you're right that to change this method so it changes the coordinate in place without making a new copy would require making `self` a mutable borrow. Since this is a library it's probably better to offer both forms as options, though I'll have to figure out the right naming convention to use.

Reply:
After sleeping on this some, I think this is a preferable form of the function. This produces cleaner code in the majority of cases but still allows someone to make the change in place if they desired. I will make sure to have `Coordinate` derive the `Copy` trait to make it a bit easier to use though, it's a fixed size and not terrible large so implicit copies shouldn't be a problem.

```rust
let mut c = Coordinate::new(...);

// with &mut self
c.adjust_epoch(t);

// vs just re-assign
c = c.adjust_epoch(t);
```